base/strings/string.jl

```diff
@@ -60,7 +60,31 @@ This representation is often appropriate for passing strings to C.
 String(s::AbstractString) = print_to_string(s)
 String(s::Symbol) = unsafe_string(unsafe_convert(Ptr{UInt8}, s))
 
-(::Type{Vector{UInt8}})(s::String) = ccall(:jl_string_to_array, Ref{Vector{UInt8}}, (Any,), s)
+unsafe_wrap(::Type{Vector{UInt8}}, s::String) = ccall(:jl_string_to_array, Ref{Vector{UInt8}}, (Any,), s)
+
+struct StringBytes <: AbstractVector{UInt8}
+    s::String
+end
+
+sizeof(s::StringBytes) = sizeof(s.s)
+length(s::StringBytes) = sizeof(s.s)
+size(s::StringBytes) = (length(s),)
+@propagate_inbounds getindex(s::StringBytes, i::Int) = codeunit(s.s, i)
+IndexStyle(::Type{StringBytes}) = IndexLinear()
+start(s::StringBytes) = 1
+next(s::StringBytes, i) = (@_propagate_inbounds_meta; (s[i], i+1))
+done(s::StringBytes, i) = (@_inline_meta; i == length(s)+1)
+
+copy(s::StringBytes) = copyto!(Vector{UInt8}(uninitialized, length(s)), s)
```

Comment:
Why are we copying immutable data? Also, seems to return the wrong type.

Reply:
Yes, this should probably be a method of `Vector{UInt8}`. Though there's some precedent for returning a different type e.g. when copying views.